workflow/lib/shared/target_utils.py

```diff
@@ -100,3 +115,251 @@ def output_to_input(output_path, wildcard_values, wildcards, ancient_output=Fals
+                if cycle_df.empty:
+                    continue
+                
+                # Base combination dict
+                base_combo = {
+                    'plate': plate,
+                    'well': well,
+                    'cycle': cycle
+                }
+                
+                # Add channel if it exists for this specific cycle
+                if 'channel' in df.columns:
+                    channels_in_cycle = cycle_df['channel'].unique()
+                    if len(channels_in_cycle) > 0:
+                        for channel in channels_in_cycle:
+                            combo = base_combo.copy()
+                            combo['channel'] = channel
+                            valid_combinations.append(combo)
+                    else:
+                        # If no channels for this cycle, just add the cycle info
+                        valid_combinations.append(base_combo)
+                else:
+                    valid_combinations.append(base_combo)
+        
+        else:  # Phenotype data
+            # Base combination dict
+            base_combo = {
+                'plate': plate,
+                'well': well
+            }
+            
+            # For phenotype data, check for missing channels only if channels are present
+            if 'channel' in df.columns:
+                all_channels = sorted(df['channel'].unique())
+                channels_in_well = sorted(well_df['channel'].unique())
+                missing_channels = set(all_channels) - set(channels_in_well)
+                
+                if missing_channels:
+                    warnings.append(f"Warning: Plate {plate}, Well {well} is missing channels: {sorted(missing_channels)}")
+                
+                for _, row in well_df.iterrows():
+                    combo = base_combo.copy()
+                    combo['channel'] = row['channel']
+                    valid_combinations.append(combo)
+            else:
+                # No channels, just add plate and well
+                valid_combinations.append(base_combo)
+    
+    return valid_combinations, warnings
+
+
+def get_sbs_combinations(df):
+    """Get valid SBS combinations and identify missing data.
+
+    Args:
+        df (pd.DataFrame): DataFrame containing SBS data.
+
+    Returns:
+        tuple: (valid_combinations, warnings)
+    """
+    valid_combinations, warnings = get_valid_combinations(df, "sbs")
+
+    for warning in warnings:
+        print(warning)
+
+    return valid_combinations
+
+
+def get_phenotype_combinations(df):
+    """Get valid phenotype combinations and identify missing data.
+
+    Args:
+        df (pd.DataFrame): DataFrame containing phenotype data.
+
+    Returns:
+        tuple: (valid_combinations, warnings)
+    """
+    valid_combinations, warnings = get_valid_combinations(df, "phenotype")
+
+    for warning in warnings:
+        print(warning)
+
+    return valid_combinations
```

Comment:
Are these functions necessary to get combinations and show warnings? Maybe we can just print warnings from the get_valid_combinations functions? I'd prefer to not have extra functions if the same thing is happening.